File: rust/src/s0001.rs
use std::collections::HashMap;
pub struct Solution {}

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut ans: Vec<i32> = Vec::with_capacity(2);
        let mut map: HashMap<i32, i32> = HashMap::new();
        for i in 0..nums.len() {
            match map.get_key_value(&(target - nums[i])) {
                Some((_, index)) => {
                    ans.push(i as i32);
                    ans.push(*index);
                }
                None => {
                    map.entry(nums[i]).or_insert(i as i32);
                }
            }
        }

        ans
    }
    pub fn two_sum_(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map: HashMap<i32, i32> = HashMap::new();
        for (i, num) in nums.iter().enumerate() {
            if map.contains_key(num) {
                return vec![i as i32, map[num]];
            }
            map.insert(target - num, i as i32);
        }
        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let mut ans = Solution::two_sum(vec![2, 7, 11, 15], 9);
        ans.sort();
        assert_eq!(ans, [0, 1]);
        let mut ans = Solution::two_sum(vec![3, 3], 6);
        ans.sort();
        assert_eq!(ans, [0, 1]);
        let mut ans = Solution::two_sum_(vec![2, 7, 11, 15], 9);
        ans.sort();
        assert_eq!(ans, [0, 1]);
        let mut ans = Solution::two_sum_(vec![3, 2, 4], 6);
        ans.sort();
        assert_eq!(ans, [1, 2]);
        let mut ans = Solution::two_sum_(vec![3, 3], 6);
        ans.sort();
        assert_eq!(ans, [0, 1]);
    }
}

File: cpp/s0001.cc
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  vector<int> twoSum(vector<int> &nums, int target) {
    std::vector<int> v = {0, 0};
    for (int i = 0; i < (int)nums.size() - 1; ++i) {
      for (int j = i + 1; j < (int)nums.size(); ++j) {
        if (nums[i] + nums[j] == target) {
          v[0] = i;
          v[1] = j;
          return v;
        }
      }
    }
    return v;
  }
  vector<int> twoSum_map(vector<int> &nums, int target) {
    std::vector<int> v = {0, 0};
    std::unordered_map<int, int> t;
    for (int i = 0; i < (int)nums.size(); ++i) {
      if (t.count(target - nums[i])) {
        v[0] = i;
        v[1] = t[target - nums[i]];
        return v;
      } else {
        t[nums[i]] = i;
      }
    }
    return v;
  }
};

int main() {
  std::vector<int> v = {2, 7, 11, 15};

  Solution s;

  cout << '[' << ' ';
  for (auto i : s.twoSum(v, 9))
    cout << i << ' ';
  cout << ']' << '\n';

  return 0;
}
